src/karts/ghost_kart.cpp

```diff
@@ -130,9 +135,45 @@ void GhostKart::update(int ticks)
         m_all_physic_info[idx].m_steer, m_all_physic_info[idx].m_speed,
         /*lean*/0.0f, idx);
 
+    // Attachment
+    // Uses custom format so that replay files remain valid if
+    // internal numbering changes.
+
+    int attach_ticks = 9000;
+
+    if ( m_all_bonus_info[idx    ].m_attachment == 0)
+        m_attachment->clear();
+    else if ( m_all_bonus_info[idx    ].m_attachment == 1 &&
+              m_attachment->getType() != Attachment::ATTACH_PARACHUTE)
+        m_attachment->set(Attachment::ATTACH_PARACHUTE,attach_ticks);
+    else if ( m_all_bonus_info[idx    ].m_attachment== 2 &&
+              m_attachment->getType() != Attachment::ATTACH_ANVIL)
+        m_attachment->set(Attachment::ATTACH_ANVIL,attach_ticks);
+    else if ( m_all_bonus_info[idx    ].m_attachment == 3 &&
+              m_attachment->getType() != Attachment::ATTACH_BOMB)
+        m_attachment->set(Attachment::ATTACH_BOMB,attach_ticks);
+    else if ( m_all_bonus_info[idx    ].m_attachment == 4 &&
+              m_attachment->getType() != Attachment::ATTACH_SWATTER)
+        m_attachment->set(Attachment::ATTACH_SWATTER,attach_ticks);
+    else if ( m_all_bonus_info[idx    ].m_attachment == 5 &&
+              m_attachment->getType() != Attachment::ATTACH_BUBBLEGUM_SHIELD)
+        m_attachment->set(Attachment::ATTACH_BUBBLEGUM_SHIELD,attach_ticks);
```

Comment:
Can you either change m_attachment to be proper attachment type, or alternatively at least cast the integers to the type for a comparison? That should be much much shorter and easier to understand.

Reply:
m_attachment is deliberately not an attachment type, see my answer to Auria below. Using the enum names as a basis rather than their raw values makes things work semantically rather than assuming a specific encoding, and avoid having old replays broken in new version if that encoding change.

That's actually the whole point of using enums with pretty names (I remember getting some flak for using raw int in an early PR).

The key difference is that the encoding for the whole code is defined by what's in the enum at compilation time, while for the replays it depends on the executable binaries at save time. So while you don't need extra-care inside the code base, when interacting with files which may have been produced with other binaries, it is a requirement.

Think of those times where you opened a text file only to discover it was broken because your editor assumed the wrong encoding (and that's best case, worse case someone did save broken values after having opened the document and received encoding errors ; breaking it completely).

Also, sure, I could default to having the number used the same as the values from the current enum, but it wouldn't be safe unless I'd keep that section of (admittedly not pretty) elseifs. And if we're stuck with them, setting "NOTHING" to "0" is better than setting "NOTHING" to "9" (current value in the enum). And using 0 to 5 is better than 0 to 3, 6 and 9. I can use a different encoding, and it does not break because the interface between the two is semantical.

Reply:
First, it's always time trial so you can only see anvil and parachute.

And if you add it an attachment for ghost kart you will have translucent kart with solid attachment mesh?

Reply:
It's currently always time trial, but there is nothing inherently preventing to expand replays in the future to normal races. Not a priority, certainly, but a real possibility.

Yes, the translucent kart with solid attachment is currently an issue. If you know how to make the attachment translucent too, tell me, it'll save me some time.

Reply:
in attachment constructor if ghost kart
    m_node = irr_driver->addAnimatedMesh(
                         attachment_manager->getMesh(Attachment::ATTACH_BOMB), "bomb",
                         NULL, std::make_shared<RenderInfo>(0.0f, true));